iwyu.cc

```diff
@@ -1436,17 +1436,18 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
       return false;
 
     // Check if the the author is not #including the file with the
-    // definition.  PublicHeaderIntendsToProvide has exactly the
-    // semantics we want.  Note if there's no definition anywhere, we
+    // definition. Note if there's no definition anywhere, we
     // say the author does not want the full type (which is a good
     // thing, since there isn't one!)
     if (const NamedDecl* dfn = GetTagDefinition(decl)) {
       if (IsBeforeInSameFile(dfn, use_loc))
         return false;
-      if (preprocessor_info().PublicHeaderIntendsToProvide(
-              GetFileEntry(use_loc), GetFileEntry(dfn))) {
+      const IwyuFileInfo* use_file =
+          preprocessor_info().FileInfoFor(GetFileEntry(use_loc));
+      if (!use_file)  // Not sure whether this is possible.
```

Comment:
It's tempting to leave this out. I think it can only happen in two cases:

* built-in types (should be covered by https://github.com/include-what-you-use/include-what-you-use/pull/1503/files#diff-7ca107c8b836c1765e8b735b74f06833bec48f308a76017886cee1d5736002d9R1412)
* precompiled headers (should be impossible as of https://github.com/include-what-you-use/include-what-you-use/commit/4f4bc53060670f095bd8192e32ffc4e0ed9272d5)

but it seems better to exit silently than crash since we don't run IWYU on a large corpus of real C++ code yet.

Should it really be `return true;`, though? Assuming it's a builtin-like thing, `false` seems potentially better.

Reply:
I just wanted to keep the old behavior when `PublicHeaderIntendsToProvide` should return `false` on nullopt, which would cause `CodeAuthorWantsJustAForwardDeclare` to return `true`.